File: core-android/src/main/java/com/ambientlink/core/Throttle.kt
package com.ambientlink.core

/**
 * Per-key leading-edge rate gate. The first [allow] for a key passes; subsequent
 * calls within [intervalMillis] are dropped. Direct analogue of Cosmo's frame
 * gate (`if tsMillis - lastFrameAt < FRAME_PROCESS_INTERVAL_MS return`) and the
 * Go backpressure.Throttle.
 *
 * Use it to drop intermediate high-rate frames (camera frames, dictation
 * partials) while letting meaningful edges through; call [reset] on a boundary
 * (capture start / new turn) so the next frame is never delayed. An interval
 * <= 0 disables throttling.
 */
class Throttle(private val intervalMillis: Long) {
    private val last = HashMap<String, Long>()

    @Synchronized
    fun allow(key: String, atMillis: Long = System.currentTimeMillis()): Boolean {
        if (intervalMillis <= 0) return true
        val prev = last[key]
        if (prev != null && atMillis - prev < intervalMillis) return false
        last[key] = atMillis
        return true
    }

    @Synchronized
    fun reset(key: String) {
        last.remove(key)
    }
}

File: host/internal/backpressure/throttle.go
package backpressure

import (
	"sync"
	"time"
)

// Throttle is a per-key leading-edge rate gate. The first Allow for a key
// passes; subsequent Allows within Interval are dropped. This is the direct
// analogue of Cosmo's frame gate:
//
//	if (tsMillis - lastFrameAt < FRAME_PROCESS_INTERVAL_MS) return  // drop
//
// Use it to drop intermediate high-rate frames (camera frames, dictation
// partials) while letting the meaningful edges through (caller resets the key
// on begin/commit so the next turn's first frame is never delayed).
//
// Safe for concurrent use. An Interval <= 0 disables throttling (Allow always
// returns true), so callers can wire it in unconditionally.
type Throttle struct {
	mu       sync.Mutex
	interval time.Duration
	last     map[string]time.Time
}

// NewThrottle returns a Throttle with the given minimum interval between
// allowed events per key.
func NewThrottle(interval time.Duration) *Throttle {
	return &Throttle{interval: interval, last: make(map[string]time.Time)}
}

// Allow reports whether an event for key at time `at` may pass. When it returns
// true it records `at` as the last-allowed time for key.
func (t *Throttle) Allow(key string, at time.Time) bool {
	if t == nil || t.interval <= 0 {
		return true
	}
	t.mu.Lock()
	defer t.mu.Unlock()
	last, ok := t.last[key]
	if ok && at.Sub(last) < t.interval {
		return false
	}
	t.last[key] = at
	return true
}

// Reset forgets key so its next Allow passes immediately. Call on a logical
// boundary (e.g. a new dictation turn or capture start).
func (t *Throttle) Reset(key string) {
	if t == nil {
		return
	}
	t.mu.Lock()
	delete(t.last, key)
	t.mu.Unlock()
}
